license_manager/service/license.go

```diff
@@ -0,0 +1,141 @@
+package service
+
+import (
+	"strings"
+	"time"
+
+	lmpb "github.com/enfabrica/enkit/license_manager/proto"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// license manages allocations and queued invocations for a single license
+// type.
+type license struct {
+	totalAvailable int                    // Constant total number of licenses available for invocations.
+	queue          []*invocation          // List of invocations waiting for a license, in FIFO order.
+	allocations    map[string]*invocation // Map of invocation ID to invocation data for an allocated license.
+}
+
+// formatLicenseType returns a unique string for a particular vendor/feature
+// combination.
+func formatLicenseType(l *lmpb.License) string {
+	return strings.Join([]string{l.GetVendor(), l.GetFeature()}, "::")
+}
+
+// Enqueue puts the supplied invocation at the back of the queue.
+func (l *license) Enqueue(inv *invocation) {
+	l.queue = append(l.queue, inv)
+}
+
+// Allocate attempts to associate the supplied invocation with a license, if
+// one is available. Returns whether a license was successfully allocated.
+func (l *license) Allocate(inv *invocation) bool {
+	if len(l.allocations) >= l.totalAvailable {
+		return false
+	}
+	l.allocations[inv.ID] = inv
+	return true
+}
+
+// Promote attempts to promote queued requests to allocations until either no
+// licenses remain or no queued requests remain.
+func (l *license) Promote() {
+	numFree := l.totalAvailable - len(l.allocations)
+	numAllocated := 0
+	for i := 0; i < numFree && i < len(l.queue); i++ {
+		l.allocations[l.queue[i].ID] = l.queue[i]
+		numAllocated++
+	}
+	l.queue = l.queue[numAllocated:]
+}
+
+// GetAllocated returns an invocation by ID if the invocation is allocated a
+// license, or nil otherwise.
+func (l *license) GetAllocated(invID string) *invocation {
+	inv, ok := l.allocations[invID]
```

Comment:
I thought that go maps, if used in the one return value form, returned the zero value for the type by default in case it was not found? Nil for pointers.  Can this be simplified with:
```
return l.allocations[invID]
```
?